[emzm2023] iP

emzm2023 opened the pull request:
### ChatbotBro

> “Your mind is for having ideas, not holding them.” – David Allen ([source] (https://dansilvestre.com/productivity-quotes))

ChatbotBro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ _SUPER FAST_ to use

All you need to do is,
1. download it.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 👍 

And it is **FREE**!

Features:

 - [  ] Managing tasks
 - [  ]Managing deadlines (coming soon)
 - [  ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
renamed chatbot, implemented greet and exit

Commit:
implemented echo and readInput methods.

Commit:
implemented add and listTasks functionality.

Commit:
renamed Duke class to Chatbot, improved functionality for adding tasks, implemented Task class and ability to mark tasks as done/undone.

Commit:
added tracking for Event, Deadline, and Todo Tasks.

Commit:
edited code in Chatbot.java and implemented semi-automated testing.

Commit:
updated functionality of Chatbot and implemented error handling with exceptions.

Commit:
implemented delete task feature, updated error handling to check for blank inputs.

Commit:
Edit comments according to coding standard

Commit:
Fixed bugs and implemented save function for tasks

Commit:
Merge branch 'branch-Level-7'

Commit:
Implemented deadline and event dates feature, fixed bugs

Commit:
Merge branch 'branch-Level-8'

Commit:
Restructured code to implement OOP principles

Commit:
Organised classes into Packages

Commit:
Merge branch 'add-gradle-support'

Commit:
Implemented JUnit testing for parseAdd and taskRep methods

Commit:
Merge branch 'junit-tests'

Commit:
Edited build.gradle for creation of JAR File

Commit:
Edited JavaDoc comments

Commit:
Implemented Find function

Commit:
Fix merge conflicts

Weiennn commented on src/main/java/chatbot/Parser.java:
```diff
@@ -0,0 +1,184 @@
+     * @throws InvalidEventException If format of event entered is invalid.
+     * @throws InvalidCommandException If format of command entered is invalid.
+     */
+    public List<String> parseAdd(String input) throws InvalidTodoException,
```
can the name of the method be changed to sound more like a verb? Maybe something like parseAddString

rngzhi replied:
```suggestion
public List<String> parseAddCommand(String input) {
```

Weiennn commented on src/main/java/chatbot/Parser.java:
```diff
@@ -0,0 +1,184 @@
+                if (description.isBlank() || start.isBlank() || end.isBlank()) {
+                    throw new InvalidEventException();
+                } else {
+                    arr.add(0, "event");
```
I like how u renamed the string inputs into something more human readable, etc description, start and end

Weiennn commented on src/main/java/chatbot/Parser.java:
```diff
@@ -0,0 +1,184 @@
+     * @throws InvalidTaskNumberException If there is no task with the task number in the taskList.
+     * @throws InvalidCommandException If format of command entered is invalid.
+     */
+    public int parseMarkTaskStatusTrue(String input) throws InvalidTaskNumberException, InvalidCommandException {
```
I like how descriptive this method name is

rngzhi replied:
Perhaps your method names could be polished to a shorter one? Such a long name is somewhat tedious. Same issues occur in the following code.

Weiennn commented on src/test/java/chatbot/ParserTest.java:
```diff
@@ -0,0 +1,73 @@
+public class ParserTest {
+
+    @Test
+    public void parseAdd_invalidCommand_exceptionThrown() {
```
I like your test names its very descriptive

Weiennn reviewed (commented):
I really like how clean and understandable your code is!!

rngzhi commented on src/main/java/chatbot/task/Event.java:
```diff
@@ -0,0 +1,48 @@
+
+public class Event extends Task {
+    /** Start time of Event. */
+    protected LocalDate start;
+    /** End time of Event. */
+    protected LocalDate end;
+
```
I like how you define the scope of variables

rngzhi commented on src/main/java/chatbot/task/Task.java:
```diff
@@ -0,0 +1,44 @@
+    protected String description;
+
+    /** The done status of the task. */
+    protected boolean isDone;
```
I like how you define variable name of boolean values

rngzhi commented on src/test/java/chatbot/TaskListTest.java:
```diff
@@ -0,0 +1,40 @@
+public class TaskListTest {
+
+    @Test
+    public void taskRep_emptyTaskList_exceptionThrown() {
```
Perhaps the name test method should follow this format, "AAATest", where AAA represents the corresponding method name.

rngzhi reviewed (commented):
Overall, you do a good job! Some method name may be too tedious and needs to be simplified.

Commit:
Implement GUI

Commit:
Refactor code through extraction of checking input methods

Parser and Storage classes have some repeated lines of code (eg.
checking if String input is blank and throwing an exception).

The common behaviours throughout the classes cause code duplication.

Extracting the common behaviour into a separate method in the respective
class helps to consolidate duplicate sections.

Let's add a new method in the Parser and Storage classes that moves
the duplicated code outside of the expression.

Commit:
Add assertions and reorganise exception handling

Exception handling for some methods accessing the chatbot TaskList
are disorganised and spread across various classes (eg. Chatbot, TaskList).

The throwing of namely the InvalidTaskNumberException in different methods is
messy and reduces code readability.

Organising that exception handling will adhere more to OOP and will be less messy.

Let's move the checking of the task number input to the TaskList class and have the methods
there throw the InvalidTaskNumberException. Let's add assertions in the classes to assert that
subsequent calls to methods in the program flow will have valid inputs after handling this
exception in the TaskList methods.

emzm2023 committed:
Merge pull request #1 from emzm2023/branch-A-Assertions

Add assertions and reorganise exception handling

Commit:
Merge branch 'master' into branch-A-CodeQuality

emzm2023 committed:
Merge pull request #2 from emzm2023/branch-A-CodeQuality

Refactor code through extraction of checking input methods

Commit:
Implement C-Tagging feature

Commit:
Merge branch 'branch-C-Tagging'

Commit:
Refactor methods through extraction for code quality

Commit:
Add tests and edit error handling for parseAdd method

Commit:
Fix bug in Ui delete confirmation printing

Commit:
Edit GUI DialogBox

Commit:
Reformat code after configuring code style

Commit:
Add User Guide and Ui.png to docs

Commit:
Fix README display issue

Commit:
Fix formatting issues with GFMD

Commit:
Update build.gradle

Commit:
Update shadowJar file name

Commit:
Fix bug with Storage loading tag for tasks

Commit:
Reformat code for bug changes

Commit:
Update README

Commit:
Implement exit for GUI

Commit:
Update README
